File: frontend/src/api/shares.ts
import type { WorkspaceRole } from './admin';
import { api } from './client';

/**
 * Suite sharing — grant/list/update/revoke per-user access to a suite, plus the directory search
 * that turns an email/name into the `user_id` a share keys on.
 */

/** Grantable share levels — `view`/`edit` only. NOT `admin` (the workspace-admin,
 *  implicit on every suite, never granted — ADR 0027) nor `owner` (the creator). */
export type SharePermission = 'view' | 'edit';

/** The caller's effective level on a suite (`SuiteRead.my_permission`): a grantable
 *  level, or `admin` (workspace-admin) / `owner` (creator). */
export type EffectivePermission = SharePermission | 'admin' | 'owner';

/** Mirrors the backend `ShareRead` — a share enriched with the grantee's identity. */
export interface Share {
  suite_id: string;
  user_id: string;
  permission: SharePermission;
  email: string;
  display_name: string | null;
}

/** Mirrors the backend `UserSummary` — the directory-picker sliver of a user. */
export interface UserSummary {
  id: string;
  email: string;
  display_name: string | null;
  /** The user's EFFECTIVE workspace role (ADR 0033). */
  role: WorkspaceRole;
}

export async function listShares(suiteId: string): Promise<Share[]> {
  const { data } = await api.get<Share[]>(`/suites/${suiteId}/shares`);
  return data;
}

export async function grantShare(
  suiteId: string,
  payload: { user_id: string; permission: SharePermission },
): Promise<Share> {
  const { data } = await api.post<Share>(`/suites/${suiteId}/shares`, payload);
  return data;
}

export async function updateShare(
  suiteId: string,
  userId: string,
  permission: SharePermission,
): Promise<Share> {
  const { data } = await api.patch<Share>(`/suites/${suiteId}/shares/${userId}`, { permission });
  return data;
}

export async function revokeShare(suiteId: string, userId: string): Promise<void> {
  await api.delete(`/suites/${suiteId}/shares/${userId}`);
}

/** Search the user directory by email/display-name substring (min 2 chars). */
export async function searchUsers(q: string): Promise<UserSummary[]> {
  const { data } = await api.get<UserSummary[]>('/users/search', { params: { q } });
  return data;
}

File: backend/app/api/v1/shares.py
"""Suite share endpoints — manage who can access a suite and at what level.

Nested under the suite. Managing shares requires `admin` on the suite (owner
qualifies); listing requires `view`. Authorization + the 404-hide / 403 split
live in `suite_authz`; this layer just wires the current user as the actor.
"""

from __future__ import annotations

import uuid
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, status
from pydantic import ConfigDict
from sqlalchemy.orm import Session

from backend.app.api.v1._base import ApiModel
from backend.app.core.auth import get_current_user
from backend.app.db.models import Share, User
from backend.app.db.session import get_db
from backend.app.services import share_service as svc

router = APIRouter(tags=["shares"])

# Grantable share levels. NOT 'owner' (the immutable creator, not a share) and
# NOT 'admin' — admin is the workspace-admin, implicit on every suite, never
# granted to a normal user (ADR 0027 / #482).
SharePermission = Literal["view", "edit"]


class ShareCreate(ApiModel):
    user_id: uuid.UUID
    permission: SharePermission


class ShareUpdate(ApiModel):
    permission: SharePermission


class ShareRead(ApiModel):
    model_config = ConfigDict(from_attributes=True)

    suite_id: uuid.UUID
    user_id: uuid.UUID
    permission: str
    # The grantee's directory identity, joined from `Share.user` so the sharing
    # UI can name collaborators without a second lookup per row.
    email: str
    display_name: str | None

    @classmethod
    def from_share(cls, share: Share) -> ShareRead:
        return cls(
            suite_id=share.suite_id,
            user_id=share.user_id,
            permission=share.permission,
            email=share.user.email,
            display_name=share.user.display_name,
        )


@router.post(
    "/suites/{suite_id}/shares",
    response_model=ShareRead,
    status_code=status.HTTP_201_CREATED,
    summary="Share a suite with a user",
)
def grant_share(
    suite_id: uuid.UUID,
    payload: ShareCreate,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> ShareRead:
    share = svc.grant_share(
        db,
        suite_id,
        actor_id=current_user.id,
        target_user_id=payload.user_id,
        permission=payload.permission,
    )
    return ShareRead.from_share(share)


@router.get(
    "/suites/{suite_id}/shares",
    response_model=list[ShareRead],
    summary="List a suite's shares",
)
def list_shares(
    suite_id: uuid.UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> list[ShareRead]:
    return [
        ShareRead.from_share(s) for s in svc.list_shares(db, suite_id, actor_id=current_user.id)
    ]


@router.patch(
    "/suites/{suite_id}/shares/{user_id}",
    response_model=ShareRead,
    summary="Change a user's permission on a suite",
)
def update_share(
    suite_id: uuid.UUID,
    user_id: uuid.UUID,
    payload: ShareUpdate,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> ShareRead:
    share = svc.update_share(
        db, suite_id, user_id, actor_id=current_user.id, permission=payload.permission
    )
    return ShareRead.from_share(share)


@router.delete(
    "/suites/{suite_id}/shares/{user_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Revoke a user's share",
)
def revoke_share(
    suite_id: uuid.UUID,
    user_id: uuid.UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> None:
    svc.revoke_share(db, suite_id, user_id, actor_id=current_user.id)
